### PR TITLE
Model parallel with bucketing

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
In this PR:
* Model parallel is modified to train complete sentences instead of just words
* Applied bucketing to model parallel, where a model is unrolled for each bucket size, and executor is generated with shared bind 
* Bucket.io is also modified to support both data parallel and model parallel 

@mli @harouwu

### CODE REVIEW

**Comment by Godricly** [comment]
Is the label_buffer missing in BucketSentenceIter for Model parallelism case?

**Comment by eric-haibin-lin** [comment]
@Godricly labels are implicitly initialized in [lstm.py](https://github.com/eric-haibin-lin/mxnet/blob/773196d28bfdbe6f1983ddf74b308328e2dcebc3/example/model-parallel-lstm/lstm.py#L251) during training

**Comment by eric-haibin-lin** [comment]
Update: added an article about model parallel with bucketing under `/how_to`.

File: docs/how_to/model_parallel_lstm.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+<img width="517" alt="screen shot 2016-05-06 at 10 13 16 pm" src="https://cloud.githubusercontent.com/assets/5545640/15089697/d6f4fca0-13d7-11e6-9331-7f94fcc7b4c6.png">
+
+In the figure above, we assign different lstm model to different GPUs. After GPU1 finish computing layer 1 with first sentence. The output will be given to GPU 2. At the same time, GPU 1 will fetch the next sentence and start training. This is significantly different from data parallelism that there's no contention to update the shared model at the end of each iteration, and most of the communication happens during pipelining intermediate results between GPU's. 
+
+In the current implementation, the layers are defined in [lstm_unroll()](https://github.com/eric-haibin-lin/mxnet/blob/master/example/model-parallel-lstm/lstm.py#L187).
```
Comment (pluskid) [inline docs/how_to/model_parallel_lstm.md]: Can you change the url to the main repo?

**Comment by pluskid** [comment]
LGTM

File: docs/how_to/model_parallel_lstm.md (hunk 1 above)
Comment (eric-haibin-lin) [inline docs/how_to/model_parallel_lstm.md]: Sure. Just updated it to main repo. Should be good to go.

**Comment by mli** [comment]
merged, thanks!

**Comment by harouwu** [comment]
Cool! Thanks Haibin, Mu and Chiyuan!

**Pull request merged**